src/notify_snmp.c

```diff
@@ -0,0 +1,868 @@
+        value = (char *)meta->nm_value.nm_string;
+      }
+    } break;
+    case NOTIFY_SNMP_VALUE_OID: {
+      notification_meta_t *meta =
+          notify_snmp_notification_meta_get(n->meta, "CurrentValue");
+      if (meta != NULL && meta->type == NM_TYPE_DOUBLE) {
+        status =
+            ssnprintf(buffer, sizeof(buffer), "%f", meta->nm_value.nm_double);
+        if ((status < 0) || ((size_t)status >= sizeof(buffer))) {
+          WARNING("notify_snmp plugin: notify_snmp_sendsnmp:"
+                  " truncate notification CurrentValue.");
+        }
+        value = buffer;
+      }
+    } break;
+    case NOTIFY_SNMP_MESSAGE_OID:
+      value = n->message;
+      break;
+    default:
+      add_value = 0;
+      break;
+    }
+
+    if (add_value) {
+      status = snmp_add_var(pdu, oids_list[i].objid, oids_list[i].len, 's',
+                            value == NULL ? "" : value);
+
+      if (status) {
+        char *errstr = NULL;
+
+        snmp_sess_error(target->sess_handle, NULL, NULL, &errstr);
+        ERROR("notify_snmp plugin: target %s: snmp_add_var"
+              " for %s (%s) failed: %s",
+              target->name, notify_snmp_oids_map_id2name(oids_list[i].id),
+              oids_list[i].string,
+              (errstr == NULL) ? "Unknown problem" : errstr);
+        snmp_free_pdu(pdu);
+        notify_snmp_exit_session(target);
+        return -1;
+      }
+    }
+  }
+
+  status = snmp_sess_send(target->sess_handle, pdu);
+  if (status == 0) {
+    char *errstr = NULL;
+
+    snmp_sess_error(target->sess_handle, NULL, NULL, &errstr);
+    ERROR("notify_snmp plugin: target %s: snmp_sess_send failed: %s.",
+          target->name, (errstr == NULL) ? "Unknown problem" : errstr);
+    snmp_free_pdu(pdu);
+    notify_snmp_exit_session(target);
+    return -1;
+  }
+
+  notify_snmp_exit_session(target);
+
+  return 0;
+}
+
+static int notify_snmp_notification(const notification_t *n,
+                                    user_data_t __attribute__((unused)) *
+                                        user_data) {
+  notify_snmp_target_t *target = notify_snmp_targets;
+  int ok = 0;
+  int fail = 0;
+
+  while (target != NULL) {
+    pthread_mutex_lock(&target->session_lock);
+    int status = notify_snmp_sendsnmp(target, (notification_t *)n);
+    pthread_mutex_unlock(&target->session_lock);
+
+    if (status < 0) {
+      fail++;
+    } else {
+      ok++;
+    }
+    target = target->next;
+  }
+
+  if (ok == 0 && fail > 0) {
+    return -1;
+  }
+
+  return 0;
+}
+
+static int notify_snmp_init(void) {
+  call_snmp_init_once();
+  return 0;
+}
+
+static int notify_snmp_shutdown(void) {
+  notify_snmp_target_t *target = notify_snmp_targets;
+  notify_snmp_oids_t *oid = notify_snmp_oids;
+
+  notify_snmp_targets = NULL;
+  while (target != NULL) {
+    notify_snmp_target_t *next;
+
+    next = target->next;
+    pthread_mutex_lock(&target->session_lock);
+    pthread_mutex_unlock(&target->session_lock);
```

Comment:
Just thinking: Is it always safe to only lock and unlock mutex here? What if there are more notifications waiting to obtain the lock? Maybe it would be better to
lock
free_target
unlock
destroy
sfree(target) //moved from free_target()